scripts/datastore_upgrade.py

```diff
@@ -0,0 +1,422 @@
+        process_entity(entity, datastore, ds_distributed)
+
+      last_key = entities[-1].keys()[0]
+      entities_checked += len(entities)
+
+      if time.time() > last_logged + LOG_PROGRESS_FREQUENCY:
+        logging.info("Checked {} entities".format(entities_checked))
+        last_logged = time.time()
+
+      if len(entities) < BATCH_SIZE:
+        break
+
+    except datastore_errors.Error as error:
+      logging.error("Error getting and validating batch of entities: {}".format(error))
+      status_dict[VALIDATE_ENTITIES] = str(error)
+      close_connections(zookeeper, db_ips, zk_ips, status_dict, keyname)
+      return
+    except AppScaleDBConnectionError as conn_error:
+      logging.error("Error getting and validating batch of entities: {}".format(conn_error))
+      status_dict[VALIDATE_ENTITIES] = str(conn_error)
+      close_connections(zookeeper, db_ips, zk_ips, status_dict, keyname)
+      return
+
+def get_entity_batch(last_key, datastore, batch_size):
+  """ Gets a batch of entities to operate on.
+  Args:
+    last_key: The last key from a previous query.
+    datastore: A reference to the batch datastore interface.
+    batch_size: The max number of entities retrieved in the datastore request.
+  Returns:
+    A list of entities with a limit of batch_size.
+  """
+  return datastore.range_query(APP_ENTITY_TABLE, APP_ENTITY_SCHEMA, last_key,
+    "", batch_size, start_inclusive=False)
+
+def process_entity(entity, datastore, ds_distributed):
+  """ Processes an entity by updating it if necessary and removing tombstones.
+  Args:
+    entity: The entity to process.
+    datastore: A reference to the batch datastore interface.
+    ds_distributed: A reference to the distributed datastore.
+  Returns:
+    True on success, False otherwise.
+  """
+  logging.debug("Process entity {}".format(str(entity)))
+  key = entity.keys()[0]
+  one_entity = entity[key][APP_ENTITY_SCHEMA[0]]
+  version = entity[key][APP_ENTITY_SCHEMA[1]]
+
+  app_id = entity_utils.get_prefix_from_entity_key(key)
+  validated_entity = ds_distributed.validated_result(app_id, entity)
+
+  is_tombstone = validated_entity[key][APP_ENTITY_SCHEMA[0]] == datastore_server.TOMBSTONE
+  if not validated_entity or is_tombstone:
+    return delete_entity_from_table(key, datastore)
+
+  if not (one_entity == validated_entity[key][APP_ENTITY_SCHEMA[0]]
+      and version == validated_entity[key][APP_ENTITY_SCHEMA[1]]):
+    return update_entity_in_table(key, validated_entity, datastore)
+
+  return True
+
+def update_entity_in_table(key, validated_entity, datastore):
+  """ Updates the APP_ENTITY_TABLE with the valid entity.
+  Args:
+    key: A str representing the row key to update in the table.
+    validated_entity: A validated entity which needs to be updated in place of
+    the current entity.
+    datastore: A reference to the batch datastore interface.
+  """
+  datastore.batch_put_entity(APP_ENTITY_TABLE, [key], APP_ENTITY_SCHEMA, validated_entity[key])
```

Comment:
Does the datastore batch_put return a list of keys successfully stored or true/false?